File: 剑指offer/40. 最小的k个数.py
########################计数排序法
class Solution:
    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        nums = [0] * 10000
        for a in arr:
            nums[a]+=1
        output=[]
        i = 0
        while len(output)<k: # 这个地方默写len(output)错了
            if nums[i]>0:
                for _ in range(nums[i]):
                    output.append(i)
            i+=1
        return output[:k] # 这个地方容易漏[:k]
#时间复杂度：O(N+K)       
#空间复杂度：O(K)???
https://leetcode-cn.com/problems/zui-xiao-de-kge-shu-lcof/solution/bu-pai-xu-bu-long-guo-lu-qi-miao-sha-88-by-libins/

File: 剑指offer/C++_Version/40. 最小的k个数.cpp
//方法1
class Solution {
public:
    vector<int> getLeastNumbers(vector<int>& arr, int k)
    {   
        quick_sort(arr, 0, arr.size()-1);
        vector<int> res;
        res.assign(arr.begin(), arr.begin()+k);
        return res;
    }

private:
    void quick_sort(vector<int>& arr, int l, int r)
    {
        if(l>=r) return;
        int i = l, j = r;
        while(i<j)
        {
            while(i<j && arr[j]>=arr[l]) j--;
            while(i<j && arr[i]<=arr[l]) i++;
            swap(arr[i], arr[j]);
        }
        swap(arr[l], arr[i]);
        quick_sort(arr, l, i - 1);
        quick_sort(arr, i + 1, r);
    }

};



//方法2
class Solution {
public:
    vector<int> getLeastNumbers(vector<int>& arr, int k)
    {   
        if(k>=arr.size()) 
            return arr;
        return quick_sort(arr, k, 0, arr.size()-1);
    }

private:
    vector<int> quick_sort(vector<int> &arr, int k, int l, int r)
    {
        int i=l, j=r;
        while(i < j)
        {
            while(i < j && arr[j] >= arr[l])  j--;
            while(i < j && arr[i] <= arr[l]) i++;
            swap(arr[i], arr[j]);
        }
        swap(arr[l], arr[i]);
        if(k < i) return quick_sort(arr, k, l, i-1);
        if(k > i) return quick_sort(arr, k, i+1, r);

        vector<int> res;
        res.assign(arr.begin(), arr.begin()+k);
        return res;
    }

};
